feral-lambda-runtime/src/main/scala/feral/lambda/runtime/LambdaRequest.scala

```diff
@@ -1,21 +1,61 @@
 package feral.lambda.runtime
 
-import cats.ApplicativeError
+import cats.effect.kernel.Concurrent
 import io.circe._
+import cats.syntax.all._
+import feral.lambda.{ClientContext, ClientContextClient, ClientContextEnv, CognitoIdentity}
+import io.circe.syntax.EncoderOps
+import org.http4s.circe._
+import org.http4s.{EntityDecoder, Headers, Response, headers}
+import org.typelevel.ci.CIString
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.FiniteDuration
 
-import java.time.Instant
-import org.http4s.Response
-
-case class LambdaRequest(
-    deadlineTimeInMs: Instant,
-    id: String,
-    invokedFunctionArn: String,
-    body: Json
+final class LambdaRequest(
+     val deadlineTimeInMs: FiniteDuration,
+     val id: String,
+     val invokedFunctionArn: String,
+     val identity: Option[CognitoIdentity],
+     val clientContext: Option[ClientContext],
+     val body: Json
     )
 
 object LambdaRequest {
-    def fromResponse[F[_]](response: Response[F])(implicit F: ApplicativeError[F, Throwable]): F[LambdaRequest] = {
-        ???
+    private[this] final val requestIdHeader = "Lambda-Runtime-Aws-Request-id"
+    private[this] final val invokedFunctionArnHeader = "Lambda-Runtime-Invoked-Function-Arn"
+    private[this] final val deadlineTimeHeader = "Lambda-Runtime-Deadline-Ms"
+    private[this] final val cognitoIdentityHeader = "Lambda-Runtime-Cognito-Identity"
+    private[this] final val clientContextHeader = "Lambda-Runtime-Client-Context"
+    def fromResponse[F[_]](response: Response[F])(implicit F: Concurrent[F]): F[LambdaRequest] = {
+        implicit val jsonDecoder: EntityDecoder[F, Json] = jsonDecoderIncremental
+        for { // TODO use custom header model for aws headers instead
+            id <- F.fromOption(response.headers.get(CIString(requestIdHeader)), new NoSuchElementException(requestIdHeader)).map(_.head.value)
+            invokedFunctionArn <- F.fromOption(response.headers.get(CIString(invokedFunctionArnHeader)), new NoSuchElementException(invokedFunctionArnHeader)).map(_.head.value)
+            deadlineTimeInMs <- F.fromOption(response.headers.get(CIString(deadlineTimeHeader)), new NoSuchElementException(deadlineTimeHeader)).map(_.head.value.toLong)
+            identity <- F.pure(response.headers.get(CIString(cognitoIdentityHeader)).flatMap(_.head.value.asJson.as[CognitoIdentity].toOption))
+            clientContext <- F.pure(response.headers.get(CIString(clientContextHeader)).flatMap(_.head.value.asJson.as[ClientContext].toOption))
+            body <- F.rethrow(jsonDecoder.decode(response, strict = false).value) // basically a reimplementation of response.as[Json] from dsl
```

Comment:
`response.as` is not DSL, it's a core method :)

By DSL, I very specifically meant things you need to import from `org.http4s.client.dsl._`